contracts/roles.js

```diff
@@ -0,0 +1,917 @@
+  }
+  if (instanceTickHours) {
+    if (!api.assert(typeof instanceTickHours === 'string' && api.BigNumber(instanceTickHours).isInteger() && api.BigNumber(instanceTickHours).gte(1), 'invalid instanceTickHours')) return;
+    params.instanceTickHours = instanceTickHours;
+  }
+  if (roleCreationFee) {
+    if (!api.assert(typeof roleCreationFee === 'string' && !api.BigNumber(roleCreationFee).isNaN() && api.BigNumber(roleCreationFee).gte(0), 'invalid roleCreationFee')) return;
+    params.roleCreationFee = roleCreationFee;
+  }
+  if (roleUpdateFee) {
+    if (!api.assert(typeof roleUpdateFee === 'string' && !api.BigNumber(roleUpdateFee).isNaN() && api.BigNumber(roleUpdateFee).gte(0), 'invalid roleUpdateFee')) return;
+    params.roleUpdateFee = roleUpdateFee;
+  }
+  if (maxSlots) {
+    if (!api.assert(typeof maxSlots === 'string' && api.BigNumber(maxSlots).isInteger() && api.BigNumber(maxSlots).gte(1), 'invalid maxSlots')) return;
+    params.maxSlots = api.BigNumber(maxSlots).toNumber();
+  }
+  if (maxInstancesPerBlock) {
+    if (!api.assert(typeof maxInstancesPerBlock === 'string' && api.BigNumber(maxInstancesPerBlock).isInteger() && api.BigNumber(maxInstancesPerBlock).gte(1), 'invalid maxInstancesPerBlock')) return;
+    params.maxInstancesPerBlock = api.BigNumber(maxInstancesPerBlock).toNumber();
+  }
+  if (maxRolesPerBlock) {
+    if (!api.assert(typeof maxRolesPerBlock === 'string' && api.BigNumber(maxRolesPerBlock).isInteger() && api.BigNumber(maxRolesPerBlock).gte(1), 'invalid maxRolesPerBlock')) return;
+    params.maxRolesPerBlock = api.BigNumber(maxRolesPerBlock).toNumber();
+  }
+  if (maxAccountApprovals) {
+    if (!api.assert(typeof maxAccountApprovals === 'string' && api.BigNumber(maxAccountApprovals).isInteger() && api.BigNumber(maxAccountApprovals).gte(1), 'invalid maxAccountApprovals')) return;
+    params.maxAccountApprovals = api.BigNumber(maxAccountApprovals).toNumber();
+  }
+  if (processQueryLimit) {
+    if (!api.assert(typeof processQueryLimit === 'string' && api.BigNumber(processQueryLimit).isInteger() && api.BigNumber(processQueryLimit).gte(1), 'invalid processQueryLimit')) return;
+    params.processQueryLimit = api.BigNumber(processQueryLimit).toNumber();
+  }
+  await api.db.update('params', params);
+};
+
+async function updateCandidateWeight(id, deltaApprovalWeight, deltaToken = null) {
+  const candidate = await api.db.findOne('candidates', { _id: id });
+  if (candidate) {
+    if (deltaToken) {
+      const role = await api.db.findOne('roles', { _id: candidate.roleId });
+      const inst = await api.db.findOne('instances', { _id: role.instanceId });
+      if (inst.voteToken !== deltaToken.symbol) return true;
+    }
+    candidate.approvalWeight = {
+      $numberDecimal: api.BigNumber(candidate.approvalWeight.$numberDecimal)
+        .plus(deltaApprovalWeight),
+    };
+    await api.db.update('candidates', candidate);
+
+    const role = await api.db.findOne('roles', { _id: candidate.roleId });
```

Comment:
if deltaToken is defined, then this is a redundant query (you've already looked up the role for deltaToken processing above).  It would be more efficient to re-use the earlier query results in that case, and only execute the query again here if deltaToken is not defined.